core/eth/state.go

```diff
@@ -59,7 +85,131 @@ func (cs *ChainState) GetOperatorSocket(ctx context.Context, blockNumber uint, o
 	return socket, nil
 }
 
-func getOperatorState(operatorsByQuorum core.OperatorStakes, blockNumber uint32) (*core.OperatorState, error) {
+// buildSocketMap returns a map from operatorID to socket address for the operators in the operatorsByQuorum
+func (cs *ChainState) buildSocketMap(ctx context.Context, operatorIds []core.OperatorID) error {
+	socketMap := make(map[core.OperatorID]*string)
+	for _, operatorID := range operatorIds {
+		// if the socket is already in the map, skip
+		if _, ok := socketMap[operatorID]; ok {
+			continue
+		}
+		socket, err := cs.Tx.GetOperatorSocket(ctx, operatorID)
+		if err != nil {
+			return err
+		}
+		socketMap[operatorID] = &socket
+	}
+
+	cs.socketMu.Lock()
+	for operatorID, socket := range socketMap {
+		cs.SocketMap[operatorID] = socket
+	}
+	cs.socketMu.Unlock()
+
+	return nil
+}
+
+// indexSocketMap indexes a block range of socket update events
+func (cs *ChainState) indexSocketMap(ctx context.Context) error {
+	currentBlockNumber, err := cs.GetCurrentBlockNumber()
+	if err != nil {
+		return err
+	}
+
+	registryCoordinator, err := cs.Tx.RegistryCoordinator(ctx)
+	if err != nil {
+		return err
+	}
+
+	logs, err := cs.Client.FilterLogs(ctx, ethereum.FilterQuery{
+		FromBlock: big.NewInt(int64(cs.socketPrevBlockNumber)),
+		ToBlock:   big.NewInt(int64(currentBlockNumber)),
+		Addresses: []gcommon.Address{registryCoordinator},
+		Topics: [][]gcommon.Hash{
+			{common.OperatorSocketUpdateEventSigHash},
+		},
+	})
+	if err != nil {
+		return err
+	}
+	if len(logs) == 0 {
+		return nil
+	}
+
+	var socketUpdates []*socketUpdateParams
+
+	// logs are in order of block number, so we can just iterate through them
+	for _, log := range logs {
+		tx, isPending, err := cs.Client.TransactionByHash(ctx, log.TxHash)
+		if err != nil {
+			return err
+		}
+		if isPending {
+			return fmt.Errorf("transaction pending for operator socket update event")
+		}
+
+		calldata := tx.Data()
+		rcAbi, err := abi.JSON(bytes.NewReader(common.RegistryCoordinatorAbi))
+		if err != nil {
+			return err
+		}
+		methodSig := calldata[:4]
+		method, err := rcAbi.MethodById(methodSig)
+		if err != nil {
+			return err
+		}
+		inputs, err := method.Inputs.Unpack(calldata[4:])
+		if err != nil {
+			return err
+		}
+
+		var socket string
+		if method.Name == "registerOperator" || method.Name == "registerOperatorWithChurn" {
+			socket = inputs[1].(string)
+		} else if method.Name == "updateOperatorSocket" {
+			socket = inputs[0].(string)
+		} else {
+			return fmt.Errorf("unknown method filtered for socket update event: %s", method.Name)
+		}
+		operatorID := core.OperatorID(log.Topics[1].Bytes())
```

Comment:
Does the length of `Bytes` need to be checked? `OperatorID` is expecting 32 bytes